File: tools/forge/bakeoff/src/jscad/B1_die.mjs
// B1 chamfered-die — hard-surface precision.
//
// Body: roundedCuboid gives a TRUE rounded fillet (radius 0.10) on all 12
// edges and 8 corners as a primitive — no manual filleting needed.
// Pips: spheres subtracted to leave spherical dents 0.08 deep.
// Colour: each dent is then re-filled with a slightly smaller solid cap that
// floats 0.001 inside the void, so it reads as a dark pip lining a real
// concave dent. Body and pips are written as separate STL parts and tinted
// on conversion (JSCAD's own colorize() only survives to 3MF/OBJ, not STL).
import { primitives, booleans, transforms, bake } from './lib.mjs'

const EDGE = 2.0 // cube edge
const FILLET = 0.10 // true rounded edge radius
const PIP_R = 0.22 // cutting sphere radius
const PIP_DEPTH = 0.08 // how deep the dent sinks below the face
const PIP_STEP = 0.45 // pip spacing within a face
const BODY_SEG = 24 // segments on the rounded edges
const PIP_SEG = 18 // segments on the pip spheres (rim smoothness)

// A pip sphere centred this far out from the die centre cuts exactly
// PIP_DEPTH into the face at EDGE/2.
const PIP_OUT = EDGE / 2 + (PIP_R - PIP_DEPTH)

// Face layouts, in face-local (u, v) units of PIP_STEP.
const LAYOUT = {
  1: [[0, 0]],
  2: [[-1, -1], [1, 1]],
  3: [[-1, -1], [0, 0], [1, 1]],
  4: [[-1, -1], [-1, 1], [1, -1], [1, 1]],
  5: [[-1, -1], [-1, 1], [0, 0], [1, -1], [1, 1]],
  6: [[-1, -1], [-1, 0], [-1, 1], [1, -1], [1, 0], [1, 1]]
}

// Opposite faces sum to 7: 1/6 on Z, 2/5 on X, 3/4 on Y.
const FACES = [
  { pips: 1, n: [0, 0, 1], u: [1, 0, 0], v: [0, 1, 0] },
  { pips: 6, n: [0, 0, -1], u: [1, 0, 0], v: [0, -1, 0] },
  { pips: 2, n: [1, 0, 0], u: [0, 1, 0], v: [0, 0, 1] },
  { pips: 5, n: [-1, 0, 0], u: [0, -1, 0], v: [0, 0, 1] },
  { pips: 3, n: [0, 1, 0], u: [-1, 0, 0], v: [0, 0, 1] },
  { pips: 4, n: [0, -1, 0], u: [1, 0, 0], v: [0, 0, 1] }
]

/** Centre of every pip sphere, in die-local coordinates. */
const pipCentres = () => {
  const out = []
  for (const f of FACES) {
    for (const [ou, ov] of LAYOUT[f.pips]) {
      out.push([0, 1, 2].map((i) =>
        f.n[i] * PIP_OUT + f.u[i] * ou * PIP_STEP + f.v[i] * ov * PIP_STEP))
    }
  }
  return out
}

const build = () => {
  const centres = pipCentres()
  const cutters = centres.map((c) =>
    primitives.sphere({ radius: PIP_R, center: c, segments: PIP_SEG }))

  const blank = primitives.roundedCuboid({
    size: [EDGE, EDGE, EDGE], roundRadius: FILLET, segments: BODY_SEG
  })
  const body = booleans.subtract(blank, ...cutters)

  // Dark inserts: 0.001 smaller than the void they sit in, so no z-fighting.
  const faceBox = primitives.cuboid({ size: [EDGE, EDGE, EDGE] })
  const pips = centres.map((c) => booleans.intersect(
    primitives.sphere({ radius: PIP_R - 0.001, center: c, segments: PIP_SEG }),
    faceBox
  ))

  const stand = (g) => transforms.translateZ(EDGE / 2, g) // sit on z = 0
  return [
    { geom: stand(body), part: 'body', color: '214,206,186' },
    { geom: stand(booleans.union(...pips)), part: 'pips', color: '38,34,40' }
  ]
}

bake('B1_die', build)

File: tools/forge/bakeoff/src/cadquery/B1_die.py
# Copyright 2026 The Dice Table Authors
# SPDX-License-Identifier: Apache-2.0
"""B1 chamfered-die -- hard-surface precision.

Cube edge 2.0, every edge rounded with a TRUE fillet of r 0.10 (one call,
`.edges().fillet(0.1)`; OCCT builds the 12 edge blends and the 8 corner
patches itself). 21 pips are spherical dents: a r 0.22 sphere whose centre
sits 0.14 outside the face, so it bites 0.08 deep and leaves a rim circle of
radius sqrt(0.22^2 - 0.14^2) = 0.170.

Colour: the die body and a set of "pip caps" go into the assembly as two
parts with two glTF materials. Each cap is the same sphere at r 0.219 --
0.001 smaller than the cutter -- intersected with the die, so it nests inside
its own dent without a coincident surface to z-fight.
"""

import cadquery as cq

from _forge import Part, Stopwatch, bake, stand_on_floor

EDGE = 2.0
FILLET = 0.10
PIP_R = 0.22
PIP_DEPTH = 0.08
PIP_OFFSET = 0.45  # pip centre spacing from the face centre

BODY_RGB = (0.76, 0.70, 0.58)
PIP_RGB = (0.11, 0.10, 0.12)

# Pip layouts in face-local (u, v), in units of PIP_OFFSET.
LAYOUT = {
    1: [(0, 0)],
    2: [(-1, -1), (1, 1)],
    3: [(-1, -1), (0, 0), (1, 1)],
    4: [(-1, -1), (-1, 1), (1, -1), (1, 1)],
    5: [(-1, -1), (-1, 1), (0, 0), (1, -1), (1, 1)],
    6: [(-1, -1), (-1, 0), (-1, 1), (1, -1), (1, 0), (1, 1)],
}

# (value, face normal, u axis, v axis). Opposite faces sum to 7, and 1-2-3
# run counter-clockwise about the (+,+,+) corner: a right-handed Western die.
FACES = [
    (1, (0, 0, 1), (1, 0, 0), (0, 1, 0)),
    (6, (0, 0, -1), (1, 0, 0), (0, -1, 0)),
    (2, (1, 0, 0), (0, 1, 0), (0, 0, 1)),
    (5, (-1, 0, 0), (0, -1, 0), (0, 0, 1)),
    (3, (0, 1, 0), (-1, 0, 0), (0, 0, 1)),
    (4, (0, -1, 0), (1, 0, 0), (0, 0, 1)),
]


def pip_centres():
    """Every pip's sphere centre, in die-local coordinates."""
    stand_off = EDGE / 2 + PIP_R - PIP_DEPTH
    for value, n, u, v in FACES:
        for du, dv in LAYOUT[value]:
            yield tuple(
                n[i] * stand_off + u[i] * du * PIP_OFFSET + v[i] * dv * PIP_OFFSET
                for i in range(3)
            )


def build():
    blank = cq.Workplane("XY").box(EDGE, EDGE, EDGE).edges().fillet(FILLET).val()

    centres = list(pip_centres())
    cutters = [cq.Solid.makeSphere(PIP_R, cq.Vector(c), angleDegrees1=-90) for c in centres]
    caps = [
        blank.intersect(cq.Solid.makeSphere(PIP_R - 0.001, cq.Vector(c), angleDegrees1=-90))
        for c in centres
    ]

    body = blank.cut(*cutters)
    return stand_on_floor(
        [
            Part("body", body, BODY_RGB),
            Part("pips", cq.Compound.makeCompound(caps), PIP_RGB),
        ]
    )


if __name__ == "__main__":
    with Stopwatch() as sw:
        parts = build()
    bake(
        "B1",
        "die",
        parts,
        tol=0.012,
        ang=0.5,
        route="native-glb",
        also_stl=True,
        build_seconds=sw.seconds,
        notes=(
            "TRUE rounded fillet r0.10 on all 12 edges from a single "
            ".edges().fillet(0.1) -- OCCT builds the 8 corner patches itself. "
            "Pips are r0.22 sphere dents 0.08 deep, rim radius 0.170. "
            "EXPORT HEAD-TO-HEAD (all three files kept): stock Assembly.export "
            "GLB writes one glTF primitive PER B-REP FACE = 99 meshes, all open; "
            "the same writer with RWGltf_CafWriter.SetMergeFaces(True) gives 2 "
            "meshes / 2 materials (shipped); stl2glb gives 1 mesh, watertight, "
            "no colour. watertight=false here is NOT a hole: the mesh is closed "
            "(Euler 2 body, 42 for the 21 caps) but OCCT's triangulator emits "
            "zero-area facets at sphere poles (2/sphere) and fillet corner "
            "patches (8/box) -- measured on isolated primitives. stl2glb drops "
            "them, which is the only reason the STL route scores watertight. "
            "Native normals come off the B-rep surface so the fillet reads "
            "tangent-smooth and the pip rims stay crisp with no crease guessing."
        ),
    )
